File: app/src/main/java/com/example/myapplication/model/MathResult.java
package com.example.myapplication.model;

/**
 * Unified result type for all mathematical operations.
 * Eliminates scattered try-catch(Exception) throughout fragments.
 */
public class MathResult<T> {

    private final T data;
    private final MathError error;

    private MathResult(T data, MathError error) {
        this.data = data;
        this.error = error;
    }

    public boolean isSuccess() { return error == null; }
    public T getData() { return data; }
    public MathError getError() { return error; }

    public static <T> MathResult<T> success(T data) {
        return new MathResult<>(data, null);
    }

    public static <T> MathResult<T> failure(MathError error) {
        return new MathResult<>(null, error);
    }

    public T orElse(T fallback) {
        return isSuccess() ? data : fallback;
    }
}

File: app/src/main/java/com/example/myapplication/model/MathResult.kt
package com.example.myapplication.model

/**
 * Unified result type for all mathematical operations.
 * Eliminates scattered try-catch(Exception) throughout fragments.
 */
class MathResult<T> private constructor(
    private val dataInternal: T?,
    val error: MathError?
) {
    val isSuccess: Boolean
        get() = error == null

    val data: T
        get() = dataInternal ?: throw IllegalStateException("Cannot access data of a failed MathResult")

    companion object {
        fun <T> success(data: T): MathResult<T> {
            return MathResult(data, null)
        }

        fun <T> failure(error: MathError): MathResult<T> {
            return MathResult(null, error)
        }
    }

    fun orElse(fallback: T): T {
        return if (isSuccess) data else fallback
    }
}
